### PR TITLE
Add editor field UI component

### PR DESCRIPTION
**Pull request by kienstra**
This should be ready for review when https://github.com/studiopress/genesis-custom-blocks/pull/43 is merged, as it's branched off that.

#### Changes
* Uses Rob's HTML and CSS
* Adds the main field grid:
the-new-field-grid

* The diff is less than it seems. This was branched off https://github.com/studiopress/genesis-custom-blocks/pull/43, but as soon as that's merged, it'll be [much simpler](https://github.com/studiopress/genesis-custom-blocks/compare/add/ui-block-settings...add/edit-block-ui-main).
* For GF-2529

#### Testing instructions
1. Follow steps 1-4 in https://github.com/studiopress/genesis-custom-blocks/pull/43#issue-518208441
1. Expected: it should look like this:
fields-grid
1. Expected: on clicking the field in the grid, the field panel shows, and you can delete the field:
on-th-wes
1. Expected: you can add fields and resize them: 
field-rsa
1. Expected: you can toggle to 'Inspector fields, but that grid doesn't really work. That's for a later ticket: 
ijsdf-field
1. Not in scope: moving the fields in the grid (GF-2528)

### CODE REVIEW

**Comment by kienstra** [comment]
You can now add and delete fields, though it only deletes the first field at the moment:

can-bls-ield

**Comment by kienstra** [comment]
## Question about icons

Hi @RobStino and @lukecarbis,
Hope you had a great weekend.

Should each of the [19 field types](https://developer.wpengine.com/genesis-custom-blocks/) have an icon?

the-field-icons

We could probably use the 'text' icon:

the-field-icons

...for other fields like 'email'. But the existing field icons don't apply as well to some fields like 'post' and 'user'.

Maybe we could use some of the block icons for the fields:

block-icons

Thanks!

**Comment by kienstra** [comment]
Also, Gutenberg has [a lot of icons](https://github.com/WordPress/gutenberg/tree/98df2666801544d8ef21c515017e21d63f299e6f/packages/icons/src/library) that we could use.

**Comment by kienstra** [comment]
The field grid is mainly working, though there are a lot of rough edges:

the-field-grid-here

**Comment by lukecarbis** [comment]
@kienstra – yes, each field type should have an icon. I imagine limiting ourselves to the Gutenberg set might be difficult, there's not a ton there. From memory, the SVGs we already have in the plugin are [Material Icons](https://material.io/resources/icons/). We could borrow from there.

**Comment by lukecarbis** [comment]
Here's a list of icons I think we should use. Note that the "style" is Outline.

Checkbox: [check_circle](https://material.io/resources/icons/?icon=check_circle&style=outline)
Color: [palette](https://material.io/resources/icons/?icon=palette&style=outline)
Email: [email](https://material.io/resources/icons/?icon=email&style=outline)
Image: [insert_photo](https://material.io/resources/icons/?icon=insert_photo&style=outline)
Multiselect: [list_alt](https://material.io/resources/icons/?icon=list_alt&style=outline)
Number: [looks_one](https://material.io/resources/icons/?icon=looks_one&style=outline)
Radio: [radio_button_checked](https://material.io/resources/icons/?icon=radio_button_checked&style=outline)
Range: [linear_scale](https://material.io/resources/icons/?icon=linear_scale&style=outline)
Select: [list](https://material.io/resources/icons/?icon=list&style=outline)
Text: [text_format](https://material.io/resources/icons/?icon=text_format&style=outline)
Textarea: [font_download](https://material.io/resources/icons/?icon=font_download&style=outline)
Toggle: [toggle_on](https://material.io/resources/icons/?icon=toggle_on&style=outline)
Url: [link](https://material.io/resources/icons/?icon=link&style=outline)
ClassicText: [elderly](https://material.io/resources/icons/?icon=elderly&style=outline)
Post: [article](https://material.io/resources/icons/?icon=article&style=outline)
Repeater: [repeat](https://material.io/resources/icons/?icon=repeat&style=outline)
RichText: [format_italic](https://material.io/resources/icons/?icon=format_italic&style=outline)
Taxonomy: [category](https://material.io/resources/icons/?icon=category&style=outline)
User: [perm_identity](https://material.io/resources/icons/?icon=perm_identity&style=outline)

@RobStino could you please review and approve?

**Comment by kienstra** [comment]
> From memory, the SVGs we already have in the plugin are Material Icons. We could borrow from there.

Thanks, that sounds good.

Would we bump the GCB license to GPLv3 to be compatible with the [Material](https://github.com/google/material-design-icons) license of Apache 2.0?

**Comment by lukecarbis** [comment]
Sure!

**Comment by kienstra** [comment]
Sorry to nitpick and be legalistic 😨 

But I'm not sure if we can change the license of the plugin without the agreement of all contributors.

Of course, GPLv2 to GPLv3 is a pretty small change. Maybe Apache 2.0 and GPLv2 are compatible, I just did a few web searches.

Update: the [GPLv2 or later](https://github.com/studiopress/genesis-custom-blocks#genesis-custom-blocks) in GCB should help. I'll look at this more.

File: js/src/block-editor/components/test/fields.js [hunk 1]
```diff
@@ -2,7 +2,7 @@
  * External dependencies
  */
 import '@testing-library/jest-dom/extend-expect';
-import React from 'react';
+import * as React from 'react';
```
Comment (kienstra) [inline js/src/block-editor/components/test/fields.js:5]: Not related to this PR, but it [looks like](https://epicreact.dev/importing-react-through-the-ages) `import * as React` is better.

File: js/src/common/helpers/getFieldsAsObject.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+		accumulator[ field.name ] = {
+			...field,
+			order: index,
+		};
+
+		return accumulator;
+	}, {} );
+};
+
+export default getFieldsAsObject;
```
Comment (kienstra) [inline js/src/common/helpers/getFieldsAsObject.js:18]: This isn't used yet, but will probably help when we move fields around.

File: js/src/edit-block/components/fields-grid.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+									onClick={ selectField }
+									onKeyPress={ selectField }
+								>
+									<div
+										className="relative flex items-center w-full p-4 bg-white border border-gray-400 rounded-sm hover:border-black"
+										id={ `field-item-${ index }` }
+									>
+										<button className="field-resize w-4 absolute -ml-2 left-0 top-0 bottom-0 focus:outline-none"></button>
+										<button>
+											<svg className="fill-current h-6 w-6" viewBox="0 0 24 24"><path d="M0 0h24v24H0z" fill="none" /><path d="M5 17v2h14v-2H5zm4.5-4.2h5l.9 2.2h2.1L12.75 4h-1.5L6.5 15h2.1l.9-2.2zM12 5.98L13.87 11h-3.74L12 5.98z" /></svg>
```
Comment (kienstra) [inline js/src/edit-block/components/fields-grid.js:155]: If it's alright, we'll add different icons for each field in a later PR. Right now, all field types have the same:

same-icon

File: js/src/common/helpers/getFieldsAsArray.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	return 0;
+};
+
+/**
+ * Gets a simplified and sorted array of the fields.
+ *
+ * @param {Object} fields The fields to simplify.
+ * @return {Array} The simplified fields.
+ */
+const getFieldsAsArray = ( fields ) => {
```
Comment (kienstra) [inline js/src/common/helpers/getFieldsAsArray.js:24]: This is just renamed and refactored, it should still work the same.

**Comment by johnstonphilip** [comment]
@kienstra Should this PR be based on a different branch now that #43 is merged?

**Comment by johnstonphilip** [comment]
Working:
- Adding new fields
- Resizing fields
- Toggle to "Inspector Fields"

Not working:
- Changing the title of the block does not change the title when attempting to insert the block on a post. 
- I am not able to add new keywords if keywords already existed upon page load. I have to remove all keyword first. Typing in the box has no effect.
- Unsure if related but I see an error in the console:
```
react_devtools_backend.js:2430 Warning: React does not recognize the `forHtml` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `forhtml` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```

**Comment by kienstra** [comment]
Hi @johnstonphilip,
Thanks for reviewing this and for your suggestions. I'll apply them.

**Comment by kienstra** [comment]
>@kienstra Should this PR be based on a different branch now that #43 is merged?

Great point, https://github.com/studiopress/genesis-custom-blocks/pull/44/commits/b4cc3697e47a38e448bf7f391bfde1062036a2e5 pull in the latest from the target branch, `feature/react-ui`

**Comment by kienstra** [comment]
https://github.com/studiopress/genesis-custom-blocks/pull/47 should fix the failed unit test.

**Comment by kienstra** [comment]
Hi @johnstonphilip,
>I am not able to add new keywords if keywords already existed upon page load. I have to remove all keyword first. Typing in the box has no effect.

Thanks for testing that. It looks alright in my local, but it definitely could be an issue.

1. Could you paste the result of `wp post get <post id> --field=content` for the `genesis_custom_block` post with that issue? 
2. Is that from using these [testing instructions](https://github.com/studiopress/genesis-custom-blocks/pull/43#issue-518208441), or how did the bug happen?

**Comment by kienstra** [comment]
>Changing the title of the block does not change the title when attempting to insert the block on a post.

This is definitely an issue. I'm working on it.

**Comment by kienstra** [comment]
Hi @johnstonphilip,
>I am not able to add new keywords if keywords already existed upon page load. I have to remove all keyword first. Typing in the box has no effect.

Could you please pull the latest and see if it still exists? I'm not sure where the error's coming from, couldn't reproduce it.

Thanks!

**Comment by johnstonphilip** [comment]
@kienstra Yes it is still happening for me on latest, unfortunately!

**Comment by kienstra** [comment]
Thanks, I also see it. Working on it now.

**Comment by kienstra** [comment]
>Changing the title of the block does not change the title when attempting to insert the block on a post.

Hi @johnstonphilip,
Good catch, the JS console error is a bug, but it looks like the fix would be in Gutenberg, not here:

console-error

Not to criticize Gutenberg, this plugin has benefited hugely from its store and components. And its very existence 🤣 

The `maxLength` of keywords is `3`, so it's expected that you can't add a new one when there are 3. But you can if there are 2 (at least in my local):

th-psdb

The console error looks to be because [this uses](https://github.com/studiopress/genesis-custom-blocks/blob/add/edit-block-ui-main/js/src/edit-block/components/keywords-section.js#L37-L48
) the `maxLength` of 3:

```js
<FormTokenField
	label={ __( 'Keywords', 'genesis-custom-blocks' ) }
	value={ block.keywords }
	maxLength={ maxNumberOfKeyword }
	onChange={ handleChange }
	onInputChange={ () => {} }
	messages={ {
		added: __( 'Keyword added.', 'genesis-custom-blocks' ),
		removed: __( 'Keyword removed.', 'genesis-custom-blocks' ),
		remove: __( 'Remove keyword', 'genesis-custom-blocks' ),
	} }
/>
```

...which causes `<FormTokenField>` to [not pass](https://github.com/WordPress/gutenberg/blob/47feef0a2aaaa38b34fb3525a35cfddaf032c719/packages/components/src/form-token-field/index.js#L585-L587
) an `onChange` prop to `<TokenInput>`:

```js
if ( ! ( maxLength && value.length >= maxLength ) ) {
	props = { ...props, onChange: this.onInputChange };
}
```

...and in `<TokenInput>`, there's [an error](https://github.com/WordPress/gutenberg/blob/47feef0a2aaaa38b34fb3525a35cfddaf032c719/packages/components/src/form-token-field/token-input.js#L31) because there's no `onChange` prop:

```js
onChange( event ) {
	this.props.onChange( {
		value: event.target.value,
	} );
}
```

**Comment by johnstonphilip** [comment]
Ah! I didn't realize that the limit was 3. That explains it. Is it possible to put that there's a 3 keyword limit in the description maybe?

**Comment by johnstonphilip** [comment]
@kienstra Also nice sleuthing!

**Comment by kienstra** [comment]
Haha, thanks! Sure, I'm looking at a way to say there's a limit.

**Comment by kienstra** [comment]
Hi @johnstonphilip,
How does this look?

the-max-

Maybe not ideal, but not as confusing as before.

**Comment by johnstonphilip** [comment]
Yeah that helps for sure! Nice

**Review by johnstonphilip — APPROVED** [review]

**Comment by kienstra** [comment]
Thanks, @johnstonphilip! Eeeeeeeh!

**Pull request merged**